caffe2/opt/optimize_ideep.cc

```diff
@@ -2,42 +2,404 @@
+
+    auto convInputs = repr::nn::getInputs(convNode);
+    if (convInputs.size() < 2) {
+      LOG(WARNING) << "Invalid convolution input size";
+      continue;
+    }
+
+    auto bnInputs = repr::nn::getInputs(bnNode);
+    if (bnInputs.size() < 5) {
+      LOG(WARNING) << "Invalid batch normalization input size";
+      continue;
+    }
+
+    // When no bias, borrow BN bias
+    if (convInputs.size() < 3) {
+      no_bias = true;
+      nn->dataFlow.createEdge(bnInputs[2], convNode);
+      convInputs = repr::nn::getInputs(convNode);
+    }
+
+#define EXPOSE_TENSOR_DATA(name, index, nodes)                                 \
+    auto* name = getTensor<itensor>(getBlob(nodes[index], ws));                \
+    if (name == nullptr) {                                                     \
+      LOG(WARNING) << #name " not a IDEEP tensor";                             \
+      continue;                                                                \
+    }                                                                          \
+    itensor name##Tensor({name->get_dims(), name->get_data_type()});           \
+    name##Tensor.reorder_from(*name);                                          \
+    assert(name##Tensor.is_public_format() && #name " not with public format");\
+    auto* name##Data = static_cast<float*>(name##Tensor.get_data_handle());
+
+    EXPOSE_TENSOR_DATA(filter, 1, convInputs);
+    EXPOSE_TENSOR_DATA(biasConv, 2, convInputs);
+
+    EXPOSE_TENSOR_DATA(scale, 1, bnInputs);
+    EXPOSE_TENSOR_DATA(biasBN, 2, bnInputs);
+    EXPOSE_TENSOR_DATA(mean, 3, bnInputs);
+    EXPOSE_TENSOR_DATA(variance, 4, bnInputs);
+
+#undef EXPOSE_TENSOR_DATA
+
+    // Assume M{CHW,HWC}
+    auto chwDim = filterTensor.get_dim(1) * filterTensor.get_dim(2) *
+        filterTensor.get_dim(3);
+    for (auto c = 0; c < filterTensor.get_dim(0); ++c) {
+      float coeff =
+          scaleData[c] / std::sqrt(varianceData[c] + bn->getEpsilon());
+      for (auto i = 0; i < chwDim; ++i) {
+        filterData[c * chwDim + i] *= coeff;
+      }
+      if (no_bias) {
+        biasConvData[c] = biasBNData[c] - meanData[c] * coeff;
+      } else {
+        biasConvData[c] = biasBNData[c] + (biasConvData[c] - meanData[c]) * coeff;
+      }
+    }
+
+    filter->reorder_from(filterTensor);
+    biasConv->reorder_from(biasConvTensor);
+    nn->dataFlow.replaceNode(convOutput, bnOutput);
+
+    nn->dataFlow.deleteNode(bnNode);
+    nn->dataFlow.deleteNode(convOutput);
+
     return true;
-  };
-  auto postprocess = [](repr::NNGraph::NodeRef conv_node) {
-    auto conv = repr::nn::get<repr::Conv>(conv_node);
-    auto annotation = conv->getMutableAnnotation();
-    if (!annotation || !isa<Caffe2Annotation>(annotation)) {
-      return;
+  }
+
+  return false;
+}
+
+void fuseConvBNForIdeep(repr::NNModule* nn, caffe2::Workspace* ws) {
+  while (fuseConvBNHelperForIdeep(nn, ws)) {
+  }
+}
+
+void fuseConvSumForIdeep(repr::NNModule* nn, caffe2::Workspace* ws) {
```

Comment:
I still don't understand what's the use case for conv+sum fusion in inference. Could you explain?

Reply:
Used to fuse residual block, for example in resnet50.
I'm afraid I did not understand your question well. Could you descrbe more on your concern?
![untitled](https://user-images.githubusercontent.com/31264804/40952366-9432bef2-68ad-11e8-8616-352c63322db6.jpg)